File: src/main/java/core/procedures/bit/BitXor.java
package core.procedures.bit;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMClass;

public final class BitXor extends AFn {

  public BitXor() {
    super(new FnArgsBuilder().min(2).rest(SCMClass.BitOp.class).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "bit-xor";
  }

  @Override
  public Long apply(Object... args) {
    long result = ((Number) args[0]).longValue();
    for (int i = 1; i < args.length; i++) {
      result ^= ((Number)args[i]).longValue();
    }
    return result;
  }
}

File: src/main/java/core/procedures/bit/BitXor.kt
package core.procedures.bit

import core.procedures.AFn
import core.procedures.Arity.AtLeast
import core.scm.Type

open class BitXor : AFn<Any?, Long?>(name = "bit-xor", isPure = true, arity = AtLeast(0), restArgsType = Type.BitOp::class.java) {

    override operator fun invoke(args: Array<out Any?>) = args.fold(0L) { r, n -> r xor (n!! as Number).toLong() }
}
